helix-core/src/shellwords.rs

```diff
@@ -1,198 +1,476 @@
     ///
     /// ```rust
-    /// use helix_core::shellwords::Shellwords;
+    /// # use helix_core::shellwords::Shellwords;
     /// assert_eq!(Shellwords::from(" ").ends_with_whitespace(), true);
     /// assert_eq!(Shellwords::from(":open ").ends_with_whitespace(), true);
     /// assert_eq!(Shellwords::from(":open foo.txt ").ends_with_whitespace(), true);
     /// assert_eq!(Shellwords::from(":open").ends_with_whitespace(), false);
-    /// #[cfg(unix)]
-    /// assert_eq!(Shellwords::from(":open a\\ ").ends_with_whitespace(), false);
-    /// #[cfg(unix)]
+    /// assert_eq!(Shellwords::from(":open a\\ ").ends_with_whitespace(), true);
     /// assert_eq!(Shellwords::from(":open a\\ b.txt").ends_with_whitespace(), false);
     /// ```
+    #[inline]
     pub fn ends_with_whitespace(&self) -> bool {
-        matches!(self.state, State::OnWhitespace)
+        self.input.ends_with(' ')
+    }
+}
+
+/// An iterator over an input string which yields arguments.
+///
+/// Splits on whitespace, but respects quoted substrings (using double quotes, single quotes, or backticks).
+/// It also handles lists enclosed in square brackets, treating them as a single argument.
+#[derive(Debug, Clone, Copy)]
+pub struct Args<'a> {
+    input: &'a str,
+    bytes: &'a [u8],
+    idx: usize,
+    start: usize,
+    in_quotes: bool,
+    quote: u8,
+    in_list: bool,
+    is_finished: bool,
+}
+
+impl<'a> Args<'a> {
+    #[inline]
+    fn parse(input: &'a str) -> Self {
+        Self {
+            input,
+            bytes: input.as_bytes(),
+            idx: 0,
+            start: 0,
+            in_quotes: false,
+            quote: b'\0',
+            in_list: false,
+            is_finished: false,
+        }
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.input.is_empty()
+    }
+
+    /// Returns the args exactly as input.
+    ///
+    /// # Examples
+    /// ```
+    /// # use helix_core::shellwords::Args;
+    /// let args = Args::from(r#"sed -n "s/test t/not /p""#);
+    /// assert_eq!(r#"sed -n "s/test t/not /p""#, args.raw());
+    ///
+    /// let args = Args::from(r#"cat "file name with space.txt""#);
+    /// assert_eq!(r#"cat "file name with space.txt""#, args.raw());
+    /// ```
+    #[inline]
+    pub fn raw(&self) -> &str {
+        self.input
+    }
+
+    #[inline(always)]
+    pub const fn empty() -> Self {
+        Self {
+            input: "",
+            bytes: &[],
+            idx: 0,
+            start: 0,
+            in_quotes: false,
+            quote: b'\0',
+            in_list: false,
+            is_finished: true,
+        }
+    }
+}
+
+#[allow(clippy::copy_iterator)]
+impl<'a> Iterator for Args<'a> {
+    type Item = &'a str;
+
+    #[inline]
+    #[allow(clippy::too_many_lines)]
+    fn next(&mut self) -> Option<Self::Item> {
+        /// Walks backwords checking for `\` and counting them to determine if escaped.
+        #[inline]
+        const fn is_escaped(bytes: &[u8]) -> bool {
```

Comment:
I actually did it this was in hopes that it would be _more_ readable. Everything here is localized to just this one function; no need to search elsewhere to try to piece together how its being parsed. Its also why I am hesitant to add a subfunction for the quotes like suggested above. Its starts to split things up for I'm not sure is any gain. I would love a definitive direction to go in.

As for the escaping, there was a couple of  thoughts around not implementing this here:

1.  This would enforce this for all commands, when really only a small number of commands would use this, and can thus just be called locally in the command itself.
2. This could limit unforeseen use cases for new commands that wouldn't be able to decide to not have a command unescaped, should it be done here.

In conclusion I had thought that it was fine to just leave it to the command itself to be aware of what kind of escaping needs to happen for it to work properly.

Reply:
> This would enforce this for all commands, when really only a small number of commands would use this, and can thus just be called locally in the command itself.

I don't think that is generally correct. The backslash escaping needs to work for all commands. All commands allow escaping space with backslash `foo\ bar` (command sees `["foo bar"]`. To still be able to type a backslash at then end of a word without joining adjacent words you need to escape `foo\\ bar` (command should see `["foo\", "bar"]`). This needs to happen for every command.


Regarding splitting up the parser what I described is a (super tiny) recursive descent parser. This is among the most common ways to describe parsers. What you do here is more of a scanner approach which I personally always find very hard to keep track of. I had to check multiple times where a state is altered and its hard to build a mental model of what can happen when. With a recursive decent parser that is a lot easier since you use the language syntax (functions) to track the current state.

The backslash tracking I described is more like the scanner/state based parsing approach you use elsewhere but that is a bit special since it applies nearly to every character. It's nicer to keep parsers linear (no backtracking) since that is easier to reason about (its also faster).

Reply:
Forgive my ignorance, when would  `foo\ bar` be chosen over using `"foo bar"`? I've not come across this practice before, which is why I removed that case when making changes. Is this from an existing paradigm, or just something that helix does?

Reply:
Another issue is that adding unescaping here would dictate that a `Cow` be used, and due to lifetimes, it would either need to be stored in `Args`, in the case that it can be pre-unescaped, but this would also mean that `Args` can no longer be `Copy`, which is paramount for the ease of use for many checks in the commands currently using `count`, and if using a `Cow` as a return, it would mean that, potentially, the `count`s could allocate strings that are never used, if the same ease of use of checks are kept. To get around this, a lot more boilerplate would need to be introduced for even trivial checks. This could lead to people just `collect`ing to get around it, which sort of defeats all the work done to remove allocations.

The design goals I had in mind for choices I made were first and foremost weighed against its usability. Given the ingrained nature of commands, for the people who would contribute to add new ones, or edit existing ones, shouldn't have to worry about a potential leaky abstraction; it was the everymans command parsing that would correctly(hopefully) match first expectation of how each arg would be interpreted. Of which, in my experience, most commands aren't worried about unescaping, they just use the near literal interpretation of the command.

If unescaping every command is something that is something that is absolutely needed, then it can't be helped, but if its just some feature that is merely a theoretical target, I would really recommend not going this route until its absolutely needed for functionality.